Implement IAM task roles as a credentials provider

indiv0 opened the pull request:
This builds upon #433 and #434, and closes #432.

indiv0 commented:
Oh, also, regarding this issue. I was wondering what we want to name it. We have an `IamProvider` for [IAM roles for EC2](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html), and this would implement [IAM roles for ECS tasks](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html). I went with `IamTaskProvider`, but I'm not sure if that's the best name for it. Ideally we'd also rename `IamProvider` to match whatever name we pick from this new provider (e.g. `IamEc2Provider`), but this would entail a breaking change.

indiv0 commented:
@adimarco @matthewkmayer this PR is ready for review from a functionality standpoint, but please read my above comment prior to merging.

indiv0 commented:
Build failed due to https://github.com/rust-lang/cargo/issues/3340.

matthewkmayer commented:
boto calls the ECS task roles [ContainerProvider](https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L944) and EC2 instance toles [InstanceMetadataProvider](https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L477).  Those names seem good to me.  I'd like to say our IAM provider predates ECS container roles so we didn't have this problem before.

matthewkmayer commented on credential/src/iam_task.rs:
```diff
@@ -0,0 +1,111 @@
+
+use {AwsCredentials, CredentialsError, ProvideAwsCredentials};
+
+const AWS_CREDENTIALS_PROVIDER_IP: &'static str = "169.254.170.2";
```
👍 Better than inlining it as with the other IAM provider.

indiv0 replied:
Yeah I'm probably going to make changes to the other providers to bring them more in line with this one (refactoring, etc.).

matthewkmayer commented on credential/src/iam_task.rs:
```diff
@@ -0,0 +1,111 @@
+        Err(_) => return Err(CredentialsError::new("Couldn't parse credentials response body.")),
+    };
+
+    fn extract_string_value_from_json(json_object: &Value, key: &str) -> Result<String, CredentialsError> {
```
We haven't used an internal function before.  I'm not sure of the benefits or drawbacks of using one.  What do we get by having it inside `parse_credentials_response` ?

indiv0 replied:
Not sure what the benefits/drawbacks are either, myself, but I put it in there because logically it's pertains specifically to the code following it (which actually calls it). The other IAM provider will also call this function once I refactor that provider, so I can move it out either now or in the refactor.

indiv0 replied:
Actually I'll just move it to the crate root now.

matthewkmayer commented on credential/src/iam_task.rs:
```diff
@@ -0,0 +1,111 @@
+    let token = try!(extract_string_value_from_json(&json_object, "Token"));
+    let expiration = try!(extract_string_value_from_json(&json_object, "Expiration"));
+
+    let expiration = try!(expiration.parse());
```
We do this in the other IAM provider but I don't recall why.  Does anyone else know?  Looks like a superfluous parse.

indiv0 replied:
Because `let expiration = try!(extract_string_value_from_json(&json_object, "Expiration"));` returns a `String`, but we need a `DateTime<UTC>`.

matthewkmayer replied:
I'm not following how we convert it to `DateTime<UTC>` without passing it any data types.  Most be missing something, would appreciate a point in the right direction.

indiv0 replied:
Rust sees that we're passing `expiration` to `AwsCredentials::new`, which expects a `DateTime<UTC>`. So it works backwards from there, and assumes that what we want `parse` to return is a `DateTime<UTC>`. There's a specific term for this, IIRC, but it's slipping my mind right now.

indiv0 replied:
Yeah that's it.

matthewkmayer commented:
Looks good to me.  Rolling the renames of IAM providers into this would be great.

indiv0 commented:
Alright I'll rename them now.

indiv0 commented:
@matthewkmayer updated, with your requested changes applied. I've also updated the `AWS-CREDENTIALS.md` to reflect the changes, and added `ContainerProvider` to the providers used by `ChainProvider`, which I forgot to do previously. I've also added `ContainerProvider` as a re-exported provider to the `rusoto` crate root (because we re-export the other providers). I've also moved that internal function to the crate root so it can be used by `InstanceMetadataProvider` in the future.

matthewkmayer commented on AWS-CREDENTIALS.md:
```diff
@@ -4,7 +4,8 @@ Rusoto has the ability to source AWS access credentials in a few different ways:
 1. Environment variables via `rusoto::EnvironmentProvider` (`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`)
 2. AWS credentials file via `rusoto::ProfileProvider`
-3. IAM instance profile via `rusoto::IamProvider`
+3. IAM EC2 instance profile via `rusoto::InstanceMetadataProvider`
```
I looked at the ordering official AWS SDKs use and it appears to be ECS container profiles then EC2 instance profiles.  The [Java SDK docs](http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html) have the clearest outline of priority.  .NET and Ruby SDKs didn't mention container profiles at all in their pages on credential sourcing.

indiv0 replied:
Alright will do.

indiv0 commented:
Up to you. I'd hate to break the build, but at the same time the changes I made were trivial so I really, really doubt that the build is gonna fail.

matthewkmayer commented:
*puts on tiny cowboy hat*

indiv0 commented:
Also, don't forget to publish a new crate.

matthewkmayer commented:
I believe that will break anyone using Rusoto and the now-known-as-`InstanceMetadataProvider` type directly.  Credentials chain users will be fine.

indiv0 commented:
Correct.
